### PR TITLE
Instantaneous recombination

### LINKED ISSUE #277: Add recombination radius based instantaneous recombination
Implement this in the dirac kernel based approach

### PR DESCRIPTION
**Pull request by dschwen**
Implements kd-tree based instantaneous recombination of Frenkel pairs with a separation distance lower than a given recombination distance `r_rec` (MyTRIMRasterizer parameter).

This also changes how simulation parameters are requested from the rasterizer. Rather than adding yet another getter I consolidate everything into a `MyTRIMRasterizer::TrimParameters` struct.

Closes #277

### CODE REVIEW

File: src/other/ThreadedRecoilLoopBase.C [hunk 1]
```diff
@@ -148,5 +159,53 @@ ThreadedRecoilLoopBase::operator() (const PKARange & pka_list)
       // done with this recoil
       delete recoil;
     }
+
+    // Process instantaneous recombination of this PKA's defects
+    if (true && !_vacancy_buffer.empty())
```
Comment (snschune) [inline src/other/ThreadedRecoilLoopBase.C]: why the `true`? `if (true && !_vacancy_buffer.empty())` and `if (!_vacancy_buffer.empty())` are totally equivalent, are they not?

File: src/other/ThreadedRecoilLoopBase.C [hunk 2]
```diff
@@ -148,5 +159,53 @@ ThreadedRecoilLoopBase::operator() (const PKARange & pka_list)
...
+      const unsigned int max_leaf_size = 50; // slightly affects runtime
+      auto point_list = PointListAdaptor<MyTRIMDefectBufferItem>(_vacancy_buffer);
+      auto kd_tree = libmesh_make_unique<KDTreeType>(
+          LIBMESH_DIM, point_list,
+          nanoflann::KDTreeSingleIndexAdaptorParams(max_leaf_size));
+
+      mooseAssert(kd_tree != nullptr, "KDTree was not properly initialized.");
+      kd_tree->buildIndex();
+
+      const Real r_rec = 25.0;
```
Comment (snschune) [inline src/other/ThreadedRecoilLoopBase.C]: why is this a fixed value?

File: include/userobjects/MyTRIMRasterizer.h [hunk 3]
```diff
@@ -85,9 +85,15 @@ class MyTRIMRasterizer : public ElementUserObject
     /// the TRIM class to instantiate in the recoil loops
     TRIMModuleEnum trim_module;
 
-    ///energy cutoff below which recoils are not followed explicitly but effects are calculated analytically
+    /// energy cutoff below which recoils are not followed explicitly but effects are calculated analytically
     Real analytical_cutoff;
 
+    /// enable instantaneous recombination
+    bool recombination;
```
Comment (snschune) [inline include/userobjects/MyTRIMRasterizer.h]: where is the leading `_` => `_recombination`

File: include/userobjects/MyTRIMRasterizer.h [hunk 4]
```diff
@@ -85,9 +85,15 @@ class MyTRIMRasterizer : public ElementUserObject
...
+
+    /// recombination radius
+    Real r_rec;
```
Comment (snschune) [inline include/userobjects/MyTRIMRasterizer.h]: This answers my earlier question, disregard.

File: src/other/ThreadedRecoilLoopBase.C [hunk 5]
```diff
@@ -157,7 +157,7 @@ ThreadedRecoilLoopBase::operator() (const PKARange & pka_list)
     }
 
     // Process instantaneous recombination of this PKA's defects
-    if (true && !_vacancy_buffer.empty())
+    if (_trim_parameters.recombination && !_vacancy_buffer.empty())
```
Comment (snschune) [inline src/other/ThreadedRecoilLoopBase.C]: same here. disregard earlier comment about `if (true && ...)`

**Comment by snschune** [comment]
Add a comment in the `ThreadedRecoilLoopBase` that instantaneous recombination is performed only within cascades and not across cascades because cascade overlap is extremely unlikely within recombination timescales.

File: include/userobjects/MyTRIMRasterizer.h (hunk 3 above)
Comment (dschwen) [reply include/userobjects/MyTRIMRasterizer.h]: I'm moving away form leading underscores in simple struct members. The benefit you have in classes with member functions is not there. I'll also go over MyTRIM and strip them off again. It turns out symbols that start with an underscore and are followed by a capital letter are reserved for internal implementation details.

**Pull request merged**